pkg/grpc/api.go

```diff
@@ -539,6 +541,177 @@ func (a *api) InvokeService(ctx context.Context, in *runtimev1pb.InvokeServiceRe
+			_, childSpan := diag.StartGRPCProducerSpanChildFromParent(ctx, span)
+			// Populate W3C traceparent to cloudevent envelope
+			corID := diag.SpanContextToW3CString(childSpan.SpanContext())
+			spanMap[i] = childSpan
+
+			var envelope map[string]interface{}
+
+			envelope, err := runtimePubsub.NewCloudEvent(&runtimePubsub.CloudEvent{
+				ID:              a.id,
+				Topic:           topic,
+				DataContentType: entries[i].ContentType,
+				Data:            entries[i].Event,
+				TraceID:         corID,
+				TraceState:      traceState,
+				Pubsub:          pubsubName,
+			})
+			if err != nil {
+				err = status.Errorf(codes.InvalidArgument, messages.ErrPubsubCloudEventCreation, err.Error())
+				apiServerLogger.Debug(err)
+				closeChildSpans(ctx, err)
+				return &runtimev1pb.BulkPublishResponse{}, err
+			}
+
+			features := thepubsub.Features()
+
+			pubsub.ApplyMetadata(envelope, features, entries[i].Metadata)
+
+			entries[i].Event, err = json.Marshal(envelope)
+			if err != nil {
+				err = status.Errorf(codes.InvalidArgument, messages.ErrPubsubCloudEventsSer, topic, pubsubName, err.Error())
+				apiServerLogger.Debug(err)
+				closeChildSpans(ctx, err)
+				return &runtimev1pb.BulkPublishResponse{}, err
+			}
+		}
+	}
+
+	req := pubsub.BulkPublishRequest{
+		PubsubName: pubsubName,
+		Topic:      topic,
+		Entries:    entries,
+		Metadata:   in.Metadata,
+	}
+
+	start := time.Now()
+	res, err := a.pubsubAdapter.BulkPublish(&req)
+	elapsed := diag.ElapsedSince(start)
+	var eventsPublished int64 = 0
+
+	bulkRes := runtimev1pb.BulkPublishResponse{}
+
+	if len(res.Statuses) != 0 {
+		for _, r := range res.Statuses {
+			resEntry := runtimev1pb.BulkPublishResponseEntry{}
+			resEntry.EntryId = r.EntryId
+			if r.Error != nil {
+				resEntry.Error = r.Error.Error()
+			}
+			if r.Status == pubsub.PublishSucceeded {
+				resEntry.Status = runtimev1pb.BulkPublishResponseEntry_SUCCESS //nolint:nosnakecase
+				eventsPublished++
+			} else {
+				resEntry.Status = runtimev1pb.BulkPublishResponseEntry_FAILED //nolint:nosnakecase
+			}
+			bulkRes.Statuses = append(bulkRes.Statuses, &resEntry)
+		}
+	}
+
+	diag.DefaultComponentMonitoring.BulkPubsubEgressEvent(context.Background(), pubsubName, topic, err == nil, eventsPublished, elapsed)
+
+	if err != nil {
+		nerr := status.Errorf(codes.Internal, messages.ErrPubsubPublishMessage, topic, pubsubName, err.Error())
+		if errors.As(err, &runtimePubsub.NotAllowedError{}) {
+			nerr = status.Errorf(codes.PermissionDenied, err.Error())
+		}
+
+		if errors.As(err, &runtimePubsub.NotFoundError{}) {
+			nerr = status.Errorf(codes.NotFound, err.Error())
+		}
+		apiServerLogger.Debug(nerr)
+		closeChildSpans(ctx, nerr)
+		return &bulkRes, nerr
+	}
+
+	closeChildSpans(ctx, nil)
+	return &bulkRes, nil
+}
+
+func populateMetadataForBulkPublishEntry(reqMeta, entryMeta map[string]string) map[string]string {
+	resMeta := map[string]string{}
```

Comment:
make(map[string]string, len(entryMeta) + len(reqMeta))

Reply:
only events that exclusive to both maps will be there in the end map ... so in this case we will be allocating extra memory

Reply:
`len(entryMeta) + len(reqMeta)` is a pessimistic case, but it's the upper bound.

If you don't specify a length, Go allocates a certain capacity (it decides how much). Then when you call `append()`, it will check if there's capacity left. If there isn't, it will allocate new memory with a buffer for extra capacity, then copy over the objects. So extra allocations would happen either way.